charts/core/src/Echart/useEchart.ts

```diff
@@ -0,0 +1,386 @@
+        }
+      }
+    }),
+    [echartsInstance],
+  );
+
+  const on: EChartsInstance['on'] = useCallback(
+    withInstanceCheck((action, callback) => {
+      // Create a unique key for this handler
+      const handlerKey = `${action}-${callback.toString()}`;
+
+      // If this exact handler is already registered, skip
+      if (activeHandlers.current.has(handlerKey)) {
+        return;
+      }
+
+      switch (action) {
+        case 'zoomselect': {
+          const zoomHandler = (params: any) => {
+            const zoomSelectionEvent: ZoomSelectionEvent = {};
+
+            if (
+              params.startValue !== undefined &&
+              params.endValue !== undefined
+            ) {
+              // Handle single axis zoom
+              const axis = params.dataZoomId?.includes('y') ? 'yAxis' : 'xAxis';
+              zoomSelectionEvent[axis] = {
+                startValue: params.startValue,
+                endValue: params.endValue,
+              };
+              callback(zoomSelectionEvent);
+            } else if (params.batch) {
+              // Handle batch zoom (multiple axes)
+              params.batch.forEach((batchItem: any) => {
+                if (batchItem.dataZoomId?.includes('y')) {
+                  zoomSelectionEvent.yAxis = {
+                    startValue: batchItem.startValue,
+                    endValue: batchItem.endValue,
+                  };
+                } else {
+                  zoomSelectionEvent.xAxis = {
+                    startValue: batchItem.startValue,
+                    endValue: batchItem.endValue,
+                  };
+                }
+              });
+              callback(zoomSelectionEvent);
+            }
+          };
+
+          // Store the wrapper function so we can remove it later
+          activeHandlers.current.set(handlerKey, zoomHandler);
+          echartsInstance.on('datazoom', zoomHandler);
+          break;
+        }
+
+        default: {
+          activeHandlers.current.set(handlerKey, callback);
+          echartsInstance.on(action, callback);
+        }
+      }
+    }),
+    [echartsInstance],
+  );
+
+  useEffect(() => {
+    (async function setupChart() {
+      try {
+        setError(null);
+
+        await initializeEcharts();
+
+        if (container) {
+          const newChart = echartsCore.init(container);
+          newChart.setOption(options);
+          window.addEventListener('resize', resizeHandler);
+          setEchartsInstance(newChart);
+          setReady(true);
+        }
+      } catch (err) {
+        setReady(false);
+        console.error(err);
+        setError(
+          err instanceof Error
+            ? err
+            : new Error('Failed to initialize ECharts'),
+        );
+      }
+    })();
+
+    return () => {
+      if (echartsInstance) {
+        window.removeEventListener('resize', resizeHandler);
+        echartsInstance.dispose();
+      }
+    };
+  }, [container]);
```

Comment:
Can we add some comments on what these `useEffects` are doing? An IIFE inside an effect callback is definitely an interesting pattern 😅